sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala

```diff
@@ -159,6 +160,66 @@ object RewritePredicateSubquery extends Rule[LogicalPlan] with PredicateHelper {
           Project(p.output, Filter(newCond.get, inputPlan))
       }
 
+    // This case takes care of predicate subqueries in join conditions that are not pushed down
+    // to the children nodes by [[PushDownPredicates]].
```

Comment:
how is this guaranteed? we run this rule in a later batch after the batch containing pushdown rules?

Reply:
Yes -- pushdown rules are in `operatorOptimizationBatch` and `RewriteSubquery` batch runs after those (almost the last batch in the Optimizer)